drivers/i2c/meson/i2c.c

```diff
@@ -566,16 +598,38 @@ static inline void handle_request(void)
     }
 }
 
-/* Right now we do not do anything to handle a response timeout, even though we should
- * be. More investigation is required.
- * We observed very frequent timeout IRQs in our example system using an I2C card reader,
- * we believe the source of this problem is a misconfiguration with the clocks, and will
- * hopefully be resovled once we have an actual clock driver.
- * See https://github.com/au-ts/sddf/issues/212 for more details.
- */
 static void handle_response_timeout(void)
 {
     LOG_DRIVER("handling timeout IRQ\n");
+
+    /* A timeout IRQ notionally occurs when a device has successfully ACKed a transaction
+    *   but stops responding sometime thereafter while we are awaiting an R/W operation.
+    *   The danger is that the device has not actually died and partially receives the
+    *   transaction and we give up on it too quickly, leaving the device in a potentially
+    *   incoherent state.
+    *
+    *   Unfortunately, the datasheet describes nothing about this IRQ so it is hard to reason
+    *   about the true conditions that give rise to it. As a result, we aim to abort the
+    *   transaction gracefully and allow the client application to recover instead.
+    */
+
+    // IRQs still arrive when nothing is happening!
+    // @mattr: since the behaviour of this IRQ is unknown, this handler may be a vector for
+    //         bugs in the future. If you ever find transactions being terminated with no
+    //         clear electrical reason, inspect this!
+    if (i2c_ifState.curr_data) {
+        volatile struct i2c_regs *regs = (volatile struct i2c_regs *)i2c_regs;
+
+        // Get state
+        uint8_t curr_token = 0;
+        uint8_t bytes_read = 0;
+        bool write_error = i2c_get_error(regs, &bytes_read, &curr_token);
```

Comment:
Is it guaranteed that `curr_token` will be valid? Should we have an assert `assert(write_error)`?

Reply:
Why is the variable called `write_error`? I was under the impression timeouts could happen on non-write transactions?

Reply:
the get error function checks the error bit in the control register. According to the Datasheet the bit is ONLY set if the i2c device generates a NACK during Writing. Hence Its a write_error

Reply:
actually upon further reading the write error only occurs if the device cant accept anymore data. So I think the logic should be changed to account for that